rclcpp/src/rclcpp/parameter_events_subscriber.cpp

```diff
@@ -0,0 +1,308 @@
+
+#include <functional>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "rclcpp/parameter_events_subscriber.hpp"
+
+namespace rclcpp
+{
+
+ParameterEventsSubscriber::ParameterEventsSubscriber(
+  rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_base,
+  rclcpp::node_interfaces::NodeTopicsInterface::SharedPtr node_topics,
+  rclcpp::node_interfaces::NodeLoggingInterface::SharedPtr node_logging,
+  const rclcpp::QoS & qos)
+: node_base_(node_base),
+  node_topics_(node_topics),
+  node_logging_(node_logging),
+  qos_(qos)
+{}
+
+void
+ParameterEventsSubscriber::add_namespace_event_subscriber(const std::string & node_namespace)
+{
+  std::lock_guard<std::recursive_mutex> lock(mutex_);
+  if (std::find(subscribed_namespaces_.begin(), subscribed_namespaces_.end(),
+    node_namespace) == subscribed_namespaces_.end())
+  {
+    subscribed_namespaces_.push_back(node_namespace);
+    auto topic = join_path(node_namespace, "parameter_events");
+    RCLCPP_DEBUG(node_logging_->get_logger(), "Subscribing to topic: %s", topic.c_str());
+
+    auto event_sub = rclcpp::create_subscription<rcl_interfaces::msg::ParameterEvent>(
+      node_topics_, topic, qos_,
+      std::bind(&ParameterEventsSubscriber::event_callback, this, std::placeholders::_1));
+
+    event_subscriptions_.push_back(event_sub);
+  }
+}
+
+void
+ParameterEventsSubscriber::remove_namespace_event_subscriber(const std::string & node_namespace)
+{
+  std::lock_guard<std::recursive_mutex> lock(mutex_);
+  auto check_sub_cb = [this, &node_namespace](
+    rclcpp::Subscription<rcl_interfaces::msg::ParameterEvent>::SharedPtr sub) {
+      return sub->get_topic_name() == join_path(node_namespace, "parameter_events");
+    };
+
+  auto it = std::find_if(
+    event_subscriptions_.begin(),
+    event_subscriptions_.end(),
+    check_sub_cb);
+
+  if (it != event_subscriptions_.end()) {
+    event_subscriptions_.erase(it);
+    subscribed_namespaces_.erase(
+      std::remove(subscribed_namespaces_.begin(), subscribed_namespaces_.end(), node_namespace));
+  }
+}
+
+void
+ParameterEventsSubscriber::set_event_callback(
+  std::function<void(const rcl_interfaces::msg::ParameterEvent::SharedPtr &)> callback,
+  const std::vector<std::string> & node_namespaces)
+{
+  std::lock_guard<std::recursive_mutex> lock(mutex_);
+  remove_event_callback();
+  std::string full_namespace;
+  for (auto & ns : node_namespaces) {
+    if (ns == "") {
+      full_namespace = node_base_->get_namespace();
+    } else {
+      full_namespace = resolve_path(ns);
+    }
+    add_namespace_event_subscriber(full_namespace);
+    event_namespaces_.push_back(full_namespace);
+  }
+
+  event_callback_ = callback;
+}
+
+void
+ParameterEventsSubscriber::remove_event_callback()
+{
+  std::lock_guard<std::recursive_mutex> lock(mutex_);
+  // Clear current vector of event namespaces and remove subscriptions
+  auto temp_namespaces = event_namespaces_;
```

Comment:
nit: Using the move constructor might be better.
Prefer not using `auto`, except the type is obvious or too complicated to write.
```cpp
  std::vector<std::string> temp_namesaces(std::move(event_namespaces_));
```